rts/motoko-rts/src/stabilization.rs

```diff
@@ -0,0 +1,462 @@
+    pub fn run(mem: &'a mut M, root: Value, stable_start: u64) -> u64 {
+        let to_space = StableMemoryStream::open(stable_start);
+        let mut serialization = Serialization { mem, to_space };
+        serialization.run(root);
+        serialization.to_space.close();
+        serialization.to_space.written_length()
+    }
+
+    fn is_null(field_value: Value) -> bool {
+        unsafe {
+            debug_assert!(!moc_null_singleton().is_forwarded());
+        }
+        field_value == unsafe { moc_null_singleton() }
+    }
+
+    fn encode_null() -> StableValue {
+        STABLE_NULL_POINTER
+    }
+
+    /// Resolve the Brooks forwarding pointer of the incremental GC by considering potential
+    /// forwarding objects (`FwdPtr`) used in Cheney's algorithm for stabilization.
+    unsafe fn resolve_gc_forwarding(object: Value) -> Value {
+        let tag = Self::read_object_tag(object);
+        if tag == TAG_FWD_PTR {
+            object
+        } else {
+            object.forward()
+        }
+    }
+    /// Read the object tag by considering potential forwarding objects (`FwdPtr`).
+    unsafe fn read_object_tag(object: Value) -> Tag {
+        // Do not call `tag()` as it dereferences the Brooks forwarding pointer of the incremental GC,
+        // which does not exist for the forwarding objects (`FwdPtr`) used by the Cheney's algorithm.
+        *(object.get_ptr() as *const Tag)
+    }
+
+    fn has_non_stable_type(old_field: Value) -> bool {
+        unsafe { old_field.tag() == TAG_CLOSURE }
+    }
+}
+
+impl<'a, M: Memory + 'a> GraphCopy<Value, StableValue, u32> for Serialization<'a, M> {
+    fn get_forward_address(&self, object: Value) -> Option<StableValue> {
+        unsafe {
+            let object = Self::resolve_gc_forwarding(object);
+            let tag = Self::read_object_tag(object);
+            match tag {
+                TAG_FWD_PTR => {
+                    let new_location = (*(object.get_ptr() as *mut FwdPtr)).fwd;
+                    Some(StableValue::serialize(new_location))
+                }
+                _ => None,
+            }
+        }
+    }
+
+    fn set_forward_address(&mut self, object: Value, target: StableValue) {
+        unsafe {
+            let object = Self::resolve_gc_forwarding(object);
+            debug_assert!(object.is_obj());
+            let fwd = object.get_ptr() as *mut FwdPtr;
+            (*fwd).tag = TAG_FWD_PTR;
+            (*fwd).fwd = target.deserialize();
+        }
+    }
+
+    fn copy(&mut self, object: Value) -> StableValue {
+        unsafe {
+            let object = Self::resolve_gc_forwarding(object);
+            debug_assert!(object.is_obj());
+            let address = self.to_space.written_length();
+            serialize(self.mem, &mut self.to_space, object);
+            StableValue::from_stable_address(address)
+        }
+    }
+
+    fn scan(&mut self) {
+        scan_serialized(self, &|context, original| {
```

Comment:
I analyzed the compiled code, and Rust seems to eliminate the lambda and completely inline its body (no extra locals used for `context` and `original`. This is because the lambda is also specified as a generic type, such that the the function is specialized for this specific lambda argument...